pkg/solver/server.go

```diff
@@ -270,6 +284,111 @@ func (solverServer *solverServer) getResult(res corehttp.ResponseWriter, req *co
 	return *result, nil
 }
 
+func (solverServer *solverServer) getJobOfferResult(res corehttp.ResponseWriter, req *corehttp.Request) {
+	vars := mux.Vars(req)
+	id := vars["id"]
+
+	err := func() *http.HTTPError {
+		jobOffer, err := solverServer.store.GetJobOffer(id)
+		if err != nil {
+			log.Error().Err(err).Msgf("error loading job offer")
+			return &http.HTTPError{
+				Message:    err.Error(),
+				StatusCode: corehttp.StatusInternalServerError,
+			}
+		}
+		if jobOffer == nil {
+			return &http.HTTPError{
+				Message:    err.Error(),
+				StatusCode: corehttp.StatusNotFound,
+			}
+		}
+
+		signerAddress, err := http.CheckSignature(req)
+		if err != nil {
+			log.Error().Err(err).Msgf("error checking signature")
+			return &http.HTTPError{
+				Message:    errors.New("not authorized").Error(),
+				StatusCode: corehttp.StatusUnauthorized,
+			}
+		}
+
+		if signerAddress != jobOffer.JobCreator {
+			log.Error().Err(err).Msgf("job creator address does not match signer address")
+			return &http.HTTPError{
+				Message:    errors.New("not authorized").Error(),
+				StatusCode: corehttp.StatusUnauthorized,
+			}
+		}
+
+		// Get the directory path
+		dirPath := GetDealsFilePath(jobOffer.DealID)
+
+		// Read directory contents
+		files, err := os.ReadDir(dirPath)
+		if err != nil {
+			return &http.HTTPError{
+				Message:    fmt.Sprintf("error reading directory: %s", err.Error()),
+				StatusCode: corehttp.StatusNotFound,
+			}
+		}
+
+		// Find the first regular file
+		var targetFile os.DirEntry
+		for _, file := range files {
+			info, err := file.Info()
+			if err != nil {
+				continue
+			}
+			if info.Mode().IsRegular() {
+				targetFile = file
+				break
+			}
+		}
+
+		if targetFile == nil {
+			return &http.HTTPError{
+				Message:    "no regular files found in directory",
+				StatusCode: corehttp.StatusNotFound,
+			}
+		}
+
+		// Get the actual filename
+		filename := targetFile.Name()
+		filePath := filepath.Join(dirPath, filename)
+
+		// Open the file
+		file, err := os.Open(filePath)
+		if err != nil {
+			return &http.HTTPError{
+				Message:    err.Error(),
+				StatusCode: corehttp.StatusInternalServerError,
+			}
+		}
+		defer file.Close()
+
+		// Set appropriate headers using the actual filename
+		res.Header().Set("Content-Disposition", fmt.Sprintf("attachment; filename=%s", filename))
+		res.Header().Set("Content-Type", "application/x-tar")
+
+		// Copy the file directly to the response
+		_, err = io.Copy(res, file)
+		if err != nil {
+			return &http.HTTPError{
+				Message:    err.Error(),
+				StatusCode: corehttp.StatusInternalServerError,
+			}
+		}
+
+		return nil
+	}()
+
+	if err != nil {
+		log.Ctx(req.Context()).Error().Msgf("error for route: %s", err.Error())
+		corehttp.Error(res, err.Error(), err.StatusCode)
+	}
+}
+
```

Comment:
Could we make a helper function to handle some of this? Looks like a lot of the code is shared with `downloadFiles`.